pallets/staking/src/integrations/democracy.rs

```diff
@@ -0,0 +1,151 @@
+use crate::traits::DemocracyReferendum;
+use crate::types::{Balance, Conviction, Vote};
+use crate::{Config, Error, Pallet};
+use frame_support::defensive;
+use frame_support::dispatch::DispatchResult;
+use orml_traits::MultiCurrencyExtended;
+use pallet_democracy::traits::DemocracyHooks;
+use pallet_democracy::{AccountVote, ReferendumIndex, ReferendumInfo};
+
+pub struct StakingDemocracy<T>(sp_std::marker::PhantomData<T>);
+
+impl<T: Config> DemocracyHooks<T::AccountId, Balance> for StakingDemocracy<T>
+where
+	T::Currency: MultiCurrencyExtended<T::AccountId, Amount = i128>,
+{
+	fn on_vote(who: &T::AccountId, ref_index: ReferendumIndex, vote: AccountVote<Balance>) -> DispatchResult {
+		let position_id = if let Some(position_id) = Pallet::<T>::get_user_position_id(who)? {
+			position_id
+		} else {
+			return Ok(());
+		};
+
+		Positions::<T>::try_mutate(position_id, |maybe_position| {
+			let position = match maybe_position.as_mut() {
+				Some(position) => position,
+				None => {
+					let e = crate::Error::<T>::InconsistentState(crate::InconsistentStateError::PositionNotFound);
+					defensive!(e);
+
+					//NOTE: This is intetional, use can't recover from this state and we don't want
+					//to block voting.
+					return Ok(());
+				}
+			};
+
+			Pallet::<T>::process_votes(position_id, position)?;
+
+			let amount = vote.balance();
+			let conviction = if let AccountVote::Standard { vote, .. } = vote {
+				match vote.conviction {
+					pallet_democracy::Conviction::None => Conviction::None,
+					pallet_democracy::Conviction::Locked1x => Conviction::Locked1x,
+					pallet_democracy::Conviction::Locked2x => Conviction::Locked2x,
+					pallet_democracy::Conviction::Locked3x => Conviction::Locked3x,
+					pallet_democracy::Conviction::Locked4x => Conviction::Locked4x,
+					pallet_democracy::Conviction::Locked5x => Conviction::Locked5x,
+					pallet_democracy::Conviction::Locked6x => Conviction::Locked6x,
+				}
+			} else {
+				Conviction::default()
+			};
+
+			let staking_vote = Vote {
+				amount: amount.min(position.stake), // use only max staked amount
+				conviction,
+			};
+
+			PositionVotes::<T>::try_mutate(position_id, |voting| -> DispatchResult {
+				match voting.votes.binary_search_by_key(&ref_index, |value| value.0) {
+					Ok(idx) => {
+						let _ = sp_std::mem::replace(&mut voting.votes[idx], (ref_index, staking_vote));
+					}
+					Err(idx) => {
+						voting
+							.votes
+							.try_insert(idx, (ref_index, staking_vote))
+							.map_err(|_| Error::<T>::MaxVotesReached)?;
+					}
+				}
+				Ok(())
+			})
+		})
+	}
+
+	fn on_remove_vote(who: &T::AccountId, ref_index: ReferendumIndex) -> DispatchResult {
+		let position_id = if let Some(position_id) = Pallet::<T>::get_user_position_id(who)? {
+			position_id
+		} else {
+			return Ok(());
+		};
+
+		PositionVotes::<T>::try_mutate_exists(position_id, |value| -> DispatchResult {
+			let voting = value.as_mut().ok_or(Error::<T>::MaxVotesReached)?;
```

Comment:
InconsistentStateError::PositionNotFound

Reply:
I think `value` is never `None` thanks to `ValueQuery` type. But it should not be a problem here. Can we change the type to `OptionQuery`? If no and it's intentional, then we can use `try_mutate` and remove this line.

Reply:
`try_mutate_exists()` returns Some/None representation of item existence  in the storage no matter of `QueryType`.
This is correct, just error is wrong, I'll fix the error.